extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt

```diff
@@ -0,0 +1,42 @@
+package org.jboss.resteasy.reactive.server.runtime.kotlin
+
+import io.vertx.core.Context
+import kotlinx.coroutines.*
+import org.eclipse.microprofile.context.ManagedExecutor
+import javax.enterprise.context.ApplicationScoped
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Application wide coroutine scope. Should start and die with the rest of the application, along with child coroutines
+ * (structured concurrency).
+ *
+ * This should be the main interception point to place user specific [CoroutineContext.Element] objects. Things
+ * like activity ids, logger context propagation, tracing, exception handlers, etc.
+ */
+@ApplicationScoped
```

Comment:
`@Singleton`?

Reply:
I would prefer to keep the ApplicationScope for now.
AFAIK In the context of Quarkus applications, there is only one application scope, technically rendering them as singletons.
If for some reason there is more than one application scope, I wouldn't want them to share the same execution context, because it would break the structured concurrency principle for that application instance.
Do you see any advantage to use `Singleton` in this case?

Reply:
They are mostly the same, the only difference is @Singleton does not use a Client Proxy, while ApplicationScoped does. This means that application scoped is lazily initialized when it is invoked, while singleton is eagerly initialized when it is injected/looked up for the first time.

ApplicationScoped can also support circular references in injection.

Reply:
In this case Singleton is probably better, as there is no real need for the client proxy.

Reply:
The CDI spec is not very clear about this, but it is as you said, no clear advantage in allowing proxies at this point. Will change.
Thanks for the clarification @stuartwdouglas